salt-cloud/libvirt: enhance libvirt provider

f-andriot opened the pull request:
… 'devices/network'

Fixes #52726

### What does this PR do?
Allow salt-cloud libvirt provider to customize 'num_cpus', 'memory' and 'devices/network' configuration options.

### What issues does this PR fix or reference?
Fixes #52726

### Previous Behavior
The options mentioned aboved were ignored by libvirt provider, the cloned domain always inherited options from source domain.

### New Behavior
The options mentioned aboved are used by libvirt provider to customize cloned domains.

### Tests written?
No

### Commits signed with GPG?
No

f-andriot committed:
salt-cloud/libvirt: add support for options 'num_cpus', 'memory', and 'devices/network'

Fixes #52726

waynew commented:
Hey, thanks for the PR!

There's some issues with lint - mixed spaces and tabs apparently (should be all spaces).

Also, would it be possible to add some tests here?

f-andriot committed:
salt-cloud/libvirt: Add ability to specify libvirt virtual machine serial number

f-andriot committed:
salt-cloud/libvirt: Fix libvirt storage pool not reloading after provisionning domain

f-andriot committed:
salt-cloud/libvirt: Add ability to create supplementary virtual disks for libvirt domains

f-andriot committed:
salt-cloud/libvirt: Do not consider IP adress beginning with '169' when provisioning libvirt domains using qemu-agent (avoid connecting to unconfigured Windows network interfaces)

f-andriot committed:
salt-cloud/libvirt: fix pylynt test

f-andriot commented:
Added more commits to add more features to libvirt provider.

Now you can:
- Define number of CPU
- Define amount of RAM
- Define several network connections
- Define guest serial number (as returned by dmidecode in guest)
- Create supplementary virtual disks

Also, fix issues during libvirt provisioning:
- Libvirt pool is not properly refreshed after VM provisioning (commit 
 #6c33758 )
- Do not attempt to connect to guest IP address beginning with '169' (invalid address given to unconfigured windows network interfaces) (commit #b1aeafb )

waynew reviewed (dismissed):
Hey, this is looking great - just a couple of things:

1. We've switched to a `master` branch, so if you could change this PR to point to `master` and rebase your changes there that would be 🎉 
2. This code is lacking automated tests. To protect against regressions, code needs to have automated tests.

We  can help train you if you have desire to write tests but just lack the skill/knowledge.

Please let us know if you're interested in making those changes!

f-andriot commented:
Hello, 
OK I will rebase on the master branch.
You are right, I have absolutely no knowledge about writing automatic tests. Please, do you have documentation about it ?

waynew commented:
Unfortunately we don't yet have any good guides - if you go sign up for a time on here: https://calendar.google.com/calendar/selfsched?sstoken=UURWQ2dRZDlKU0twfGRlZmF1bHR8OWU3N2NmZjgwODY5ODVjNGU3NWRhMTRhYTI4MDM3ZmY I can meet with you either via chat or hangouts/zoom/???

waynew commented:
@f-andriot we'll be running a test clinic this Thursday during our Community Open Hour. If you haven't subscribed to our Community Events Calendar you can get it [here](https://calendar.google.com/calendar/embed?src=saltstack.com_md73c3ufcs2eqbsmmnike4em80%40group.calendar.google.com). Click the *totally intuitive* + at the very bottom right of the screen to add it to your own calendar.

waynew commented:
@f-andriot are you still interested in driving this to completion?

f-andriot commented:
> @f-andriot are you still interested in driving this to completion?

Hello, I'm still working on developing for this PR inside my company, to add more features.
I will push again here when I'm done.
But I'll probably won't have time to make the automatic test stuff.

waynew commented:
@f-andriot thanks for the update!

Let us know when you're done with the code side of things - we'll be hosting test clinics regularly, and if you're not interested or able to write the tests we'll find someone who will be able to write the tests 👍

sagetherage commented:
@f-andriot please rebase this against the master branch and update the title with brackets if it is a WIP = Work In Progress on your end, we will help with tests, but if we have not heard from you in some time we will close this PR as abandoned.

f-andriot committed:
salt-cloud/libvirt: salt-cloud/libvirt: add support 'autostart' option

f-andriot committed:
salt-cloud/libvirt: clouds/libvirt: allow to passthrough physical disks from hypervisor

f-andriot committed:
salt-cloud/libvirt: clouds/libvirt: allow to override the cloned virtual disk name, avoid generating random name

f-andriot commented:
Hello, I've rebased against master branch as requested.
But the PR is still against develop branch. Is there a way to change it ?

sagetherage commented:
@f-andriot you can do a `git-rebase --onto` git-scm has some good reference guides [here](https://git-scm.com/docs/git-rebase) and looks like you have some merge conflicts, as well.

f-andriot committed:
salt-cloud/libvirt: Use metadata preallocation when cloning/creating qcow2 virtual hard disks

f-andriot committed:
salt-cloud/libvirt: Use double-quotes instead of single-quotes

f-andriot commented:
Hello, I've added new commits and rebased against master branch as requested.
Still, the PR was done agains the 'develop' branch, that's why it looks like there are many conflicting files.
Is there a way to change the destination branch ?

waynew commented:
@f-andriot I went ahead and pointed the PR to master for you. If you have need of it in the future, all you have to do is click the "edit" at top right and then you can change where you're merging to 👍

waynew commented:
re-run debian8-py3

f-andriot committed:
salt-cloud/libvirt: appease linter

f-andriot committed:
salt-cloud/libvirt: add 'thin_provision' option to enable or disable disk preallocation

f-andriot committed:
salt-cloud/libvirt: remove python2 support and six library

f-andriot committed:
salt-cloud/libvirt: appease linter (hook 'isort')

f-andriot committed:
salt-cloud/libvirt: appease linter (hook 'black')

sagetherage commented:
@f-andriot Thank you for rebasing! looks like there is a merge conflict if you can resolve, please.

@cbosdo can you also give a review here, please?

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -441,37 +594,203 @@ def create(vm_):
+                            vol_path = pool_target + "/" + vol_name
+
+                            try:
+                                vol = pool.storageVolLookupByName(vol_name)
```
You surely want to refresh the pool here. Maybe even ensure it is started...

f-andriot replied:
OK, I've added the pool.refresh() and pool.create()

cbosdo replied:
thanks

cbosdo reviewed (changes requested):
Thanks a lot for taking care of the libvirt cloud driver. Your PR seems fairly good to me, but I left a few comments to share code or make it more robust.

f-andriot committed:
salt-cloud/libvirt: add 'usb' option to specify USB controller version (1, 2 or 3)

f-andriot committed:
salt-cloud/libvirt: the <smbios> and <sysinfo> XML tags are created if missing in source XML domain

f-andriot committed:
salt-cloud/libvirt: fix virtual NIC allocation when >10 NICs are defined

f-andriot committed:
salt-cloud/libvirt: allow to specify SCSI devices address for SCSI disks (controller:bus:target:unit)

f-andriot committed:
salt-cloud/libvirt: add the 'start' and 'stop' functions to start/stop VMs from the salt-cloud command-line

f-andriot commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -390,11 +444,110 @@ def create(vm_):
+            # Configure serial number
+            if serial:
+                entry_xml = domain_xml.find("./sysinfo/system/entry")
+                entry_xml.text = str(serial)
```
OK I've added the code; the only problem is that "get_xml_node()" creates the XML tag at the end of XML file, whereas libvirt XML requires that <sysinfo> XML tag is at a specific location

cbosdo replied:
libvirt doesn't impose any order among sibling elements. You can perfectly add the `<sysinfo></sysinfo>` after `<devices></devices>`. Just mind that libvirt may change the place of the element though... so you may have a hard time diffing the `virsh dumpxml` output with what you provided libvirt.

f-andriot replied:
Hello, that's what I though initially, but when I tried get_xml_node() directlry, it put the <sysinfo> XML tag at end of file as you describe, then I got a libvirt error regarding dictionary validation failure (in domaincommon.rng).
That's why I put the <sysinfo> on top of the<os> tag, now I have no dictionary error.
(BTW I'm using libvirt 4.5.0 from CentOS 7.9, not the latest libvirt version I guess)

f-andriot replied:
OK this was a mistake in my code, after fixing my syntax it now works as expected. No need to create sysinfo at fixed location.

cbosdo replied:
cool, thanks!

f-andriot commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -390,11 +444,110 @@ def create(vm_):
+            # Configure amount of memory
+            if memory:
+                try:
+                    memory_num, memory_unit = re.findall(
+                        r"[^\W\d_]+|\d+.\d+|\d+", memory
+                    )
+                    if memory_unit.lower() == "mb":
+                        memory_mb = int(memory_num)
+                    elif memory_unit.lower() == "gb":
+                        memory_mb = int(float(memory_num) * 1024.0)
+                    else:
+                        err_msg = "Invalid memory type specified: '{}'".format(
+                            memory_unit
+                        )
+                        log.error(err_msg)
+                        return {"Error": err_msg}
+                except (TypeError, ValueError):
+                    memory_mb = int(memory)
```
This code is copy/paste from the vmware provider :)
From what I see in _handle_unit() , it is using per-1000 units, but we want per-1024 units here. E.g. user will ususally type "1GB" for virtual memory, and it should result in 2048MB, not 2000MB.

cbosdo replied:
`_handle_unit` handles the units as defined by libvirt. So it handles both 1024 and 1000 versions. Typically `1MiB` is `1024KiB` while `1MB` is `1000KB`. Additionally `1m` will default to a 1024 multiple. This fits the usual megabyte vs [mebibyte definition](https://en.wikipedia.org/wiki/Byte#Multiple-byte_units)

f-andriot replied:
OK I will use that, and I will add this precision to the comments too.
Older people (like me) still have 1MB = 1024KB in mind,  but this is wrong according to the definition.

f-andriot replied:
Done in latest commit

f-andriot committed:
salt-cloud/libvirt: ensure the storage pool is started prior to creating virtual disks

f-andriot committed:
salt-cloud/libvirt: don't create sysinfo XML explicitely

f-andriot committed:
salt-cloud/libvirt: use _handle_memery to calculate memory amount for VM

f-andriot committed:
salt-cloud/libvirt: fix GB/GiB units in comments

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -66,6 +115,8 @@
     SaltCloudNotFound,
     SaltCloudSystemExit,
 )
+from salt.modules.virt import _handle_unit
```
I wonder if that function should be moved to `salt.utils.virt` module instead

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+        _f for _f in [cfg.get("libvirt") for cfg in providers.values()] if _f
+    ]
+    for provider in providers_to_check:
+        conn = __get_conn(provider["url"])
+        log.info("looking at %s", provider["url"])
+        try:
+            domain = conn.lookupByName(name)
+            found.append(domain)
+        except libvirtError:
+            pass
```
I would be better to close the connection in a `finally` block

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+        _f for _f in [cfg.get("libvirt") for cfg in providers.values()] if _f
+    ]
+    for provider in providers_to_check:
+        conn = __get_conn(provider["url"])
+        log.info("looking at %s", provider["url"])
+        try:
+            domain = conn.lookupByName(name)
+            found.append(domain)
+        except libvirtError:
+            pass
```
Please close the connection in a `finally` block

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+        _f for _f in [cfg.get("libvirt") for cfg in providers.values()] if _f
+    ]
+    for provider in providers_to_check:
+        conn = __get_conn(provider["url"])
+        log.info("looking at %s", provider["url"])
+        try:
+            domain = conn.lookupByName(name)
+            found.append(domain)
+        except libvirtError:
+            pass
```
Please close the connection in a `finally` block

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+    )
+
+    log.info("Starting domain %s", found[0].name())
+    found[0].create()
```
You may get a `libvirtError` here too

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+        transport=__opts__["transport"],
+    )
+
+    log.info("Stopping domain %s", found[0].name())
+    found[0].shutdown()
```
You may also get a `libvirtError` here too.

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
+        transport=__opts__["transport"],
+    )
+
+    log.info("Rebooting domain %s", found[0].name())
+    found[0].shutdown()
```
You may get a `libvirtError` here too

cbosdo reviewed (changes requested):
There are still some libvirt connection handling improvements to be made, but the rest of the libvirt handling seems good.

I don't have sufficient salt-cloud knowledge to review the global design of your PR though. In my dreams this provider would use the `virt` module... but that is another story :wink:

f-andriot committed:
salt-cloud/libvirt: add conn.close() and libVirtError to start(), stop() and reboot()

cbosdo commented on salt/cloud/clouds/libvirt.py:
```diff
@@ -1159,6 +1159,8 @@ def start(name, call=None):
         except libvirtError:
             pass
+        finally:
+            conn.close()
```
Sorry for not having been precise enough. I think closing here may be too early: I would bet the domain objects stored in `found` array would then be freed.

There are two possible solutions:
1. Store the connection handles in a list and close them once all the domains have been started (or any other action for the similar functions)
2. Close the connections that gave no result within the `for` loop and the others with something like:
```python
for dom in found:
    dom.connect().close()
```

f-andriot replied:
Hello, I will look into that :)
I'm not the original author for these start, stop and reboot functions, so I need to talk with my colleague about what behaviour is expected exactly.

From what I understand, the functions are targeting a single VM, not a list.
To locate where the VM is, we connect to each provider in a loop.
If we find zero or more than one VM (e.g. there are VM with the same name on different hypervisor), we do nothing.

The functions are working in their current states, not sure how. I agree that closing the connection before calling start() should fail, but it does not.

cbosdo replied:
OK, I see better how it works now. The real connection `close()` is done after the start so we are safe

f-andriot committed:
salt-cloud/libvirt: better handling libvirt connection in start(), stop() an reboot()

f-andriot commented:
OK so the start(), stop() and reboot() functions were copy/paste from existing destroy() function.
It looks like destroy() was already handling conn.close() , so I've done the same in our functions.

Ch3LL commented:
ping @cbosdo is this one ready on your end?

cbosdo reviewed (approved):
LGTM

Ch3LL commented:
Thanks for the review @cbosdo  Looks like the lint and pre-commit tests are failing and still need to be cleaned up @f-andriot

Ch3LL reviewed (changes requested):
this will require a changelog and test coverage

Ch3LL commented:
bump @f-andriot did you see my comments?

dwoz commented:
Closing due to lack of activity. Anyone should feel free to re-open or create a new PR if they want to see this through to the end.
